### PR TITLE
[SPARK-20723][ML]Add intermediate storage level to tree based classifiers

### PR DESCRIPTION
**Pull request by phatak-dev**
## What changes were proposed in this pull request?

Currently Random Forest implementation caches the intermediate data using MEMORY_AND_DISK storage level. This creates issues in low memory scenarios. So we should expose an expert param intermediateStorageLevel which allows user to customise the storage level. This is similar to als options specified in below jira

https://issues.apache.org/jira/browse/SPARK-14412

## How was this patch tested?
unit test cases.

### CODE REVIEW

**Comment by phatak-dev** [comment]
Current unit test case is rudimentary. Any help in improving it is appreciated.

**Comment by MLnick** [comment]
If other algos are going to use this param, we should put in in shared params - i.e. update `SharedParamsCodeGen` and have the relevant estimators inherit from the trait.

**Comment by phatak-dev** [comment]
@MLnick should i go ahead and move it to SharedParamsCodeGen?

**Comment by MLnick** [comment]
Yes let's make it shared if possible.

**Comment by phatak-dev** [comment]
I have updated the code to have the parameter in sharedParamsCodeGen.

**Comment by phatak-dev** [comment]
@MLnick can you start a jenkins build?

File: mllib/src/main/scala/org/apache/spark/ml/param/shared/sharedParams.scala [hunk 1]
```diff
@@ -406,4 +409,21 @@ private[ml] trait HasAggregationDepth extends Params {
...
+/**
+ * Trait for shared param intermediateStorageLevel (default: "MEMORY_AND_DISK").
+ */
+private[ml] trait HasIntermediateStorageLevel extends Params {
+
+  /**
+   * Param for Param for StorageLevelfor intermediate datasets.
+   * @group expertParam
+   */
+  final val intermediateStorageLevel: Param[String] = new Param[String](this, "intermediateStorageLevel", "Param for StorageLevelfor intermediate datasets", (s: String) => Try(StorageLevel.fromString(s)).isSuccess && s != "NONE")
```
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/param/shared/sharedParams.scala]: should be a space: `StorageLevelfor` -> `StorageLevel for`

File: mllib/src/test/scala/org/apache/spark/ml/tree/impl/RandomForestSuite.scala [hunk 2]
```diff
@@ -199,7 +199,7 @@ class RandomForestSuite extends SparkFunSuite with MLlibTestSparkContext {
     withClue("DecisionTree requires number of features > 0," +
       " but was given an empty features vector") {
       intercept[IllegalArgumentException] {
-        RandomForest.run(rdd, strategy, 1, "all", 42L, instr = None)
+        RandomForest.run(rdd, strategy, 1, "all", 42L, instr = None, "MEMORY_AND_DISK")
```
Comment (MLnick) [inline mllib/src/test/scala/org/apache/spark/ml/tree/impl/RandomForestSuite.scala]: extract out the storage level `MEMORY_AND_DISK` into a `val` that can be reused

File: mllib/src/test/scala/org/apache/spark/ml/classification/DecisionTreeClassifierSuite.scala [hunk 3]
```diff
@@ -398,6 +398,14 @@ class DecisionTreeClassifierSuite
 
     testDefaultReadWrite(model)
   }
+
+  test("intermediate dataset storage level") {
```
Comment (MLnick) [inline mllib/src/test/scala/org/apache/spark/ml/classification/DecisionTreeClassifierSuite.scala]: Ideally we need to test that the actual storage levels for the intermediate RDDs are correct. I did this for ALS - not sure if the same approach might be used here.

**Comment by MLnick** [comment]
Ok to test

**Comment by phatak-dev** [comment]
@MLnick i am working on fixes you recommended. Meantime jenkins build doesn't seems to be started.

**Comment by MLnick** [comment]
jenkins test this please

**Comment by phatak-dev** [comment]
There are around 33 binary incompatibility issues as I made this a shared param. Some of them from build logs are below
```scala
[error]  * abstract synthetic method org$apache$spark$ml$param$shared$HasIntermediateStorageLevel$_setter_$intermediateStorageLevel_=(org.apache.spark.ml.param.Param)Unit in trait org.apache.spark.ml.param.shared.HasIntermediateStorageLevel is inherited by class ALSParams in current version.
[error]    filter with: ProblemFilters.exclude[InheritedNewAbstractMethodProblem]("org.apache.spark.ml.param.shared.HasIntermediateStorageLevel.org$apache$spark$ml$param$shared$HasIntermediateStorageLevel$_setter_$intermediateStorageLevel_=")
[error]  * method getIntermediateStorageLevel()java.lang.String in class org.apache.spark.ml.recommendation.ALS is declared final in current version
[error]    filter with: ProblemFilters.exclude[FinalMethodProblem]("org.apache.spark.ml.recommendation.ALS.getIntermediateStorageLevel")
[error]  * method intermediateStorageLevel()org.apache.spark.ml.param.Param in class org.apache.spark.ml.recommendation.ALS is declared final in current version
[error]    filter with: ProblemFilters.exclude[FinalMethodProblem]("org.apache.spark.ml.recommendation.ALS.intermediateStorageLevel")
[error]  * abstract synthetic method org$apache$spark$ml$param$shared$HasIntermediateStorageLevel$_setter_$intermediateStorageLevel_=(org.apache.spark.ml.param.Param)Unit in trait org.apache.spark.ml.param.shared.HasIntermediateStorageLevel is inherited by class RandomForestRegressorParams in current version.
```
Should I add them to the exclude file?

**Comment by phatak-dev** [comment]
@MLnick Any updates on this?

File: mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala [hunk 4]
```diff
@@ -129,7 +129,7 @@ private[recommendation] trait ALSModelParams extends Params with HasPredictionCo
  * Common params for ALS.
  */
 private[recommendation] trait ALSParams extends ALSModelParams with HasMaxIter with HasRegParam
-  with HasPredictionCol with HasCheckpointInterval with HasSeed {
+  with HasPredictionCol with HasCheckpointInterval with HasSeed with HasIntermediateStorageLevel{
```
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: space: `HasIntermediateStorageLevel {`

**Review by WeichenXu123 — COMMENTED** [review]
add these to exclude file.
```
ProblemFilters.exclude[InheritedNewAbstractMethodProblem]("org.apache.spark.ml.param.shared.HasIntermediateStorageLevel.org$apache$spark$ml$param$shared$HasIntermediateStorageLevel$_setter_$intermediateStorageLevel_=")
ProblemFilters.exclude[FinalMethodProblem]("org.apache.spark.ml.recommendation.ALS.getIntermediateStorageLevel")
ProblemFilters.exclude[FinalMethodProblem]("org.apache.spark.ml.recommendation.ALS.intermediateStorageLevel")
```

**Comment by phatak-dev** [comment]
@WeichenXu123 Done the changes to mima.

**Comment by WeichenXu123** [comment]
Have you checked other algorithms which can also apply this parameter ?

**Comment by phatak-dev** [comment]
I have added to all the algorithms I know that uses intermediate state.

**Comment by phatak-dev** [comment]
Can you please initiate the Jenkins build?

**Comment by phatak-dev** [comment]
@WeichenXu123 resolved merge conflicts. Can you initiate jenkins build?

**Comment by rkarimi** [comment]
perhaps related: 
Big Random Forest Models (example: 100 or more trees with depth of around 20):



Big models can be trained effectively even on machines with limited RAM (such as C series in AWS). However, they fail during .transform stage. My guess is that it tries to load model to RAM of individual worker nodes which do not have that much capacity. Above proposed fix (relying on DISK as well) can potentially fix this.

An alternative fix is to load model tree by tree evaluate results and store evaluation/transform. This way, we may not need the Disk storage yet and it would better resemble the training process (also tree by tree).

**Comment by WeichenXu123** [comment]
@phatak-dev This PR is pretty old so the biggest challenge is going to be updating it to the current master branch, and this is a minor feature IMO. Shall we close it now ? Thanks!

**Comment by WeichenXu123** [comment]
@phatak-dev This PR is stale. If there's nobody interested in this and no further updates, would you mind to close it ? Thanks!

**Comment by WeichenXu123** [comment]
The PR is stale and no update for a long time. I will close this. But still appreciate your contribution ! @phatak-dev

**Pull request closed without merging**